pandas/tests/test_strings.py

```diff
@@ -2924,7 +2932,7 @@ def test_pipe_failures(self):
 
         tm.assert_series_equal(result, exp)
 
-        result = s.str.replace('|', ' ')
+        result = s.str.replace('|', ' ', regex=False)
```

Comment:
Hmm Ok. I think this is correct but arguably an API breaking change so make sure we make note of that in the whatsnew

Reply:
how is this an api change? regex=True is the default

Reply:
Was just thinking of this particular instance and any where a user was passing in a single character that may have special meaning with a regex. This previously would directly replace a pipe but now requires `regex=False` in user code, so it could cause some breakage.

Being extra conservative but not tied to the request then if you feel its over communicating.

Reply:
Exactly, but if you think this documentation is not necessary, let me know and I can change.

Reply:
Hmm. This is a potentially disruptive change...

Can we:

1. Change the default `regex=None` for `.str.replace`
2. Detect when a length-1 character is a regex symbol
3. Warn that it'll change in the future to interpret that character as a regex, not a literal
4. set `regex=False` for now to preserve the old (buggy) behavior?

Reply:
I actually would be +1 to option 4 over the long run because we get a decent amount of issues crop up due to confusion around this defaulting to a regex replace. However, I don't think that's really a viable solution w.r.t this change as it would only preserve behavior for one character patterns but break the old behavior for everything else.

This is definitely a weird one because the one character pattern through this bug essentially disregarded the `regex` argument. We could go with some combination of options 2 and 3 above, though the documented behavior of it today would imply that something like "." should match everything so I have a slight preference for the fix as is

Reply:
Sorry, I meant that list as a "sequence of steps", not a list of options :)

I was assuming that we wanted the future behavior to be "|" is a regex, manually specify `regex=False` if you don't want that. But as you say, regex=False is probably what people want here...

Reply:
I don't think it's possible to fix this without breaking something. We'd either be breaking "incorrect" (as dictated by documentation) usage of single character patterns as this change does, or we'd have to break multi-character regular expression patterns.

I'm not opposed to the latter as it would align this method with the `DataFrame` behavior, but that would require a deprecation cycle while allowing this bug to persist I think. Not sure atm if I have a strong preference for either

Reply:
I believe changing the default behavior of the regex variable goes against the default behavior of the replace function.
Today it's expected that this method assumes regex=True and my initial proposal was to adjust the behavior of the function for this premise

What I have suggested is to adjust the behavior now and leave the rewrite of this code for a future moment.

Reply:
Let's be explicit. I'd like for

```python
In [5]: pd.Series(['a.c']).str.replace('.', 'b')
Out[5]:
0    abc
dtype: object
```

to not change without warning.

https://github.com/pandas-dev/pandas/pull/24809#discussion_r248670917 suggests a path for that where we warn for a timespan, before changing to the correct behavior, of not ignoring `regex`.

Reply:
OK I agree - that's probably the best go-forward path, save the first point which I don't understand.

@alarangeiras can you raise a FutureWarning here instead?

Reply:
> save the first point which I don't understand.

Changing the default `regex=None`? That's so we can detect if we need a warning or not.

If the user passes `.str.replace('.', 'b', regex=True)`, we know to interpret the `.` as `re.compile('.')`, so the output would be `'bbb'`.

If the user passes `.str.replace('.', 'b', regex=False)`, we know that they want a literal `.`, so the output is `'abc'`.

We'll use regex=None to see if the user is explicit or not.

Reply:
Why not just warn when the length of the pattern is 1 and `regex=True`? Whether or not the user explicitly typed that or relied on the default argument they'd hit the same bug at the end of the day. Don't see value in introducing a `None` value into a True/False field currently

Reply:
> Why not just warn when the length of the pattern is 1 and regex=True? 

Then I think there would be no way to have

```python
In [5]: pd.Series(['a.c']).str.replace('.', 'b')
# Warning: Interpreting '.' as a literal, not a regex...
Out[5]:
0    abc
dtype: object
```

```python
# no warning
In [5]: pd.Series(['a.c']).str.replace('.', 'b', regex=True)
Out[5]:
0    bbb
dtype: object
```

unless I"m missing something.

Reply:
Following this line of reasoning, from what I understand, every bug found should issue a warning of future adjustment?